File: src/test/java/io/inbot/utils/AESUtilsTest.java
package io.inbot.utils;

import static org.assertj.core.api.StrictAssertions.assertThat;

import java.security.SecureRandom;
import java.util.HashSet;
import java.util.Set;
import org.testng.annotations.Test;


@Test
public class AESUtilsTest {

    public void shouldConvertToFromHex() {
        byte[] iv = new byte[20];
        new SecureRandom().nextBytes(iv);
        String hexString = AESUtils.byteArrayToHexString(iv);
        byte[] bytes = AESUtils.hexStringToByteArray(hexString);
        assertThat(bytes).isEqualTo(iv);
    }

    public void shouldEncryptAndDecryptUsingSaltAndPassword() {
        assertThat(AESUtils.decrypt("salt", "password", AESUtils.encrypt("salt", "password", "secretsecret"))).isEqualTo("secretsecret");
    }

    public void shouldEncryptAndDecryptUsing256BitKey() {
        byte[] key = new byte[32];
        new SecureRandom().nextBytes(key);
        String plainText = "secretsecret";
        String encrypted = AESUtils.encrypt(key, plainText);
        assertThat(AESUtils.decrypt(key, encrypted)).isEqualTo(plainText);
    }

    public void shouldEncryptAndDecryptUsingGenerated256BitBase64Key() {
        String plainText = "secretsecret";
        String key = AESUtils.generateAesKey();
        String encrypted = AESUtils.encrypt(key, plainText);
        assertThat(AESUtils.decrypt(key, encrypted)).isEqualTo(plainText);
    }

    public void shouldPreserveCompatibility() {
        String encrypted="B8B32CCF0471B3FA305C3DF04D879E79$5or2YFCkt50EtCnqakG4KyMRItErJyIHhflVlNZTAVLxLwSWFEHDeWfpvyZ4z8lG";
        assertThat(AESUtils.decrypt("salt", "password", encrypted)).isEqualTo("secretsecret");
    }

    public void shouldNotGenerateSameEncryptedEver() {
        Set<String> encrypted= new HashSet<>();
        for(int i=0;i<100;i++) {
            // each encrypted value is hashed
            encrypted.add(AESUtils.encrypt("salt", "password", "secretsecret"));
        }
        assertThat(encrypted.size()).isEqualTo(100);
    }

    @Test(expectedExceptions=IllegalArgumentException.class)
    public void shouldThrowExecptionForIncorrectPassword() {
        AESUtils.decrypt("salt", "passwd", AESUtils.encrypt("salt", "password", "secretsecret"));
    }

    @Test(expectedExceptions=IllegalArgumentException.class)
    public void shouldThrowExecptionForIncorrectSalt() {
        AESUtils.decrypt("slt", "password", AESUtils.encrypt("salt", "password", "secretsecret"));
    }
}

File: src/test/java/io/inbot/utils/AESUtilsTest.kt
package io.inbot.utils

import org.assertj.core.api.Assertions.assertThat

import io.inbot.utils.AESUtils
import org.assertj.core.api.StrictAssertions

import java.security.SecureRandom
import org.testng.annotations.Test


@Test
class AESUtilsTest {

    @Throws(Exception::class)
    fun shouldConvertToFromHex() {
        val iv = ByteArray(20)
        SecureRandom().nextBytes(iv)
        val hexString = AESUtils.byteArrayToHexString(iv)
        val bytes = AESUtils.hexStringToByteArray(hexString)
        StrictAssertions.assertThat(bytes).isEqualTo(iv)
    }

    fun shouldEncryptAndDecryptUsingSaltAndPassword() {
        val plainText = "secretsecret"
        val encrypted = AESUtils.encrypt("salt", "password", plainText)
        StrictAssertions.assertThat(AESUtils.decrypt("salt", "password", encrypted)).isEqualTo(plainText)
    }

    fun shouldEncryptAndDecryptUsing256BitKey() {
        val key = ByteArray(32)
        SecureRandom().nextBytes(key)
        val plainText = "secretsecret"
        val encrypted = AESUtils.encrypt(key, plainText)
        StrictAssertions.assertThat(AESUtils.decrypt(key, encrypted)).isEqualTo(plainText)
    }

    fun shouldEncryptAndDecryptUsingGenerated256BitBase64Key() {
        val plainText = "secretsecret"
        val key = AESUtils.generateAesKey()
        val encrypted = AESUtils.encrypt(key, plainText)
        StrictAssertions.assertThat(AESUtils.decrypt(key, encrypted)).isEqualTo(plainText)
    }
}
